### PR TITLE
Refactor `CLI::Parser`.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Assign global options to `Homebrew.args` and partially remove special-casing of global options.

### CODE REVIEW

File: Library/Homebrew/cli/args.rb [hunk 1]
```diff
@@ -10,7 +10,7 @@ class Args < OpenStruct
       # undefine tap to allow --tap argument
       undef tap
 
-      def initialize(argv = ARGV.freeze, set_default_args: false)
+      def initialize(argv = ARGV.dup.freeze, set_default_args: false)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/args.rb:13]: We do want to `.freeze` `ARGV` at some point (although I agree here isn't the best place).

File: Library/Homebrew/cli/parser.rb [hunk 2]
```diff
@@ -153,31 +156,54 @@ def summary
...
-          named_args = @parser.parse(argv)
-        rescue OptionParser::InvalidOption => e
-          $stderr.puts generate_help_text
-          raise e
+        i = 0
+        remaining = []
+
+        argv, non_options = split_double_dash(argv)
+
+        while i < argv.count
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cli/parser.rb:167]: What's this all doing? It's not super obvious. Does it warrant a separate function and/or comment?

File: Library/Homebrew/cli/args.rb (hunk 1 above)
Comment (reitermarkus) [reply Library/Homebrew/cli/args.rb]: It's frozen now in `#parse` by default.

File: Library/Homebrew/cli/parser.rb (hunk 2 above)
Comment (reitermarkus) [reply Library/Homebrew/cli/parser.rb:165]: Moved this into a separate method.

Comment (reitermarkus) [reply Library/Homebrew/cli/parser.rb:165]: As to what it does: Basically the same as `@parser.parse`, while being able to ignore invalid options.

**Comment by SeekingMeaning** [comment]
Perhaps `Homebrew.args` should be re-added but deprecated

**Comment by reitermarkus** [comment]
This is an internal API and I don't think anything other than `livecheck` is still using it.

**Comment by dawidd6** [comment]
Homebrew.args is not internal. It is used for parsing arguments in external commands too, like `brew livecheck` or:
https://github.com/Homebrew/homebrew-linux-dev
https://github.com/Homebrew/homebrew-services

`brew bundle` is fixed I see. There could be much more external commands, that are in use and they can be broken now like `brew livecheck` is.

**Comment by MikeMcQuaid** [comment]
> Perhaps `Homebrew.args` functionality should be re-added but deprecated

I agree. We've required code changes (or flat out broken) pretty much all the official taps with these changes and I'm pretty disappointed to see @reitermarkus not involved in most of these fixes. There's also been a tonne of regressions in behaviour. I realise that changes like this are hard to do without breakage but I think we need to work out a different approach for any future work like this.

**Comment by MikeMcQuaid** [comment]
@reitermarkus Can you check the `args` usage on the remaining taps in the Homebrew organisation: https://github.com/Homebrew?q=homebrew-&type=&language=

Thanks!

**Comment by reitermarkus** [comment]
> I'm pretty disappointed to see @reitermarkus not involved in all of these fixes.

 If issues were opened, I would have certainly fixed them myself, but I won't complain that other maintainers beat me to it by opening PRs instead.

> I think we need to work out a different approach for any future work like this.

I definitely agree. I think it's pretty much impossible to get this right if none of the official external commands are tested when `brew` changes. I didn't even know half of these command existed, to be honest.

> Can you check the `args` usage on the remaining taps in the Homebrew organisation

Seems like all are using `args` now instead of the global `Homebrew.args`, except for the still global `debug?`, `quiet?` and `verbose?`.

**Comment by MikeMcQuaid** [comment]
> Seems like all are using `args` now instead of the global `Homebrew.args`

Confirmed 👍🏻 



> I think it's pretty much impossible to get this right if none of the official external commands are tested when `brew` changes.

They are tested but the tests are pretty basic (i.e. don't run options). Running `rspec` on them all seems potentially overkill but could be investigated.

**Comment by reitermarkus** [comment]
> Running `rspec` on them all seems potentially overkill

Even running `rspec` on all of them wouldn't have cought the `args` bug, since the specs only test the `lib` part of the commands and don't run any sort of integration test.

**Pull request merged**